### DISCUSSION TITLE
Displaying huge letters without memory overflow

### DISCUSSION THREAD

**Discussion by akmere** [grovesNL/glyphon#73] [Q&A]
Hello guys,
is there any straigtforward way to display huge letters without overflowing memory? Even at the cost of quality. At some level of magnification/scaling my program is getting super laggy.
![image](https://github.com/grovesNL/glyphon/assets/21156950/b7401468-05fd-45ad-985f-7cb310bf3b4b). As of now if I want to make a text bigger, I'm just putting the bigger scale to TextArea. 
```
let text_area = TextArea {
            buffer: self.buffer.as_ref().unwrap(),
            left: lefto + x_translation,
            top: topo + (configuration.height as f32 - y_translation),
            scale: **scaling**,
            (...)
```
I can put some fake lower resolution in text_rendere.prepare(), letters will be appropriately bigger, but it surely isn't pragmatic.

I'm using the current version of from github. Thanks for all help!

**Comment by grovesNL** [comment]
Hi! 👋 Are you able to reproduce this on the hello-world example with a large font size? #74 might help slightly with startup time

**Comment by Patiga** [comment]
Hey, so far I really enjoy using this library, however I'm stumped right now with text scaling, in the same boat as @akmere .

What I want to do is render text at different sizes/zoom levels, without populating the atlas with differently sized glyphs. Best case for me would be that I can simply specify at which resolution level the glyphs go into the atlas.
I'm okay with the text looking pixelated when zooming in and getting aliasing effects when zoomed out.
Like akmere, I thought that the `scale` field of `TextArea` would do exactly that, but that fills up my atlas very quickly.

Thanks for helping and maintaining the lib! :)

**Comment by BillyDM** [comment]
For rendering large text it's probably better to just use the swash crate to render large text directly into a single bitmap, and then draw that bitmap on the GPU as a regular image. It would be nice if there was a convenient way to do this.

**Reply by Patiga** [reply]
Hm, I kind of hijacked this discussion thread, but my use-case isn't about rendering huge text.
To give my concrete example, I am reimplementing a 2D renderer for a game and am currently doing the nameplates. The size of the nameplates scale the same as the players. So when you zoom in and out, the text gradually gets smaller and bigger, completely thrashing the atlas.
The reference implementation simply has the text on one resolution, which is good for typical zoom distances and gets pixelated when you zoom in, but so does everything else in the game.

**Reply by Patiga** [reply]
So I am in need of a texture atlas, as the name plates can have arbitrary symbols of any language in them, but don't want glyphon to try to accommodate every zoom distance

**Reply by Patiga** [reply]
Maybe I should start a new discussion? My symptoms are similar, so I thought that it might be considered a duplicate ^^

**Reply by BillyDM** [reply]
Oh ok. Well glyphon is specifically designed for rendering high quality small text in GUI applications. It is not designed for games. What you might want is text rendered using SDFs (signed distance fields). That is a pretty common technique used in games.

**Reply by BillyDM** [reply]
There is this crate https://crates.io/crates/kaku

**Reply by BillyDM** [reply]
Oh wait, I think I see what you are saying now. Yeah, glyphon doesn't have a way to apply a transformation to the text (i.e. scale, rotation). It is designed to render text at a 1:1 scale for maximum sharpness with antialiasing.

I suppose it *could* be possible to add an optional transformation matrix to the shader. (Though keep in mind any transformation applied to small text will make it look blurry.)

**Reply by Patiga** [reply]
I can see how "I don't want maximum sharpness" is an odd request at first.
In my case I don't care about rotation/shearing, but only wish to convince glyphon that it can use the glyphs from the atlas, even if they are not of the perfect quality.

> There is this crate https://crates.io/crates/kaku

I happened to find that one too, but the amount of text I want to render will go up dramatically later (like up to thousands of strings), and afaik sdf rendering of text will take a big hit performance-wise.

I'm not super familiar with text rendering. What is the limitation of the size of `cosmic_text::Buffer` (also with the `cosmic_text::Metrics`)? I'm surprised that glyphon can scale the text to any size even with buffer sizes.

**Reply by BillyDM** [reply]
Ok, so the way glyphon (or more specifically swash) works, is that tells the swash library the exact pixel size and location of each glyph rendered (including 4 sub-pixel bins). Swash uses this information to not only properly antialias the text (including sub-pixel antialiasing), but to also apply "hinting", which is where the text engine intelligently snaps lines to whole pixels to make the text appear sharper (it's a complex process).

You might be able to get around the atlas issue by using a separate glyphon instance for each "scalable" text buffer.

**Reply by BillyDM** [reply]
Again, it could be possible to add an extra transformation matrix to the shader if you are ok with blurry text. Maybe something like that could be done with a feature flag?

**Reply by BillyDM** [reply]
Oh wait, but you would need padding around every glyph in the atlas for the transformation matrix to work. I can't remember if glyphon has support for that or not.

**Reply by BillyDM** [reply]
Something else you might be able to do is render the glyphon output to a separate GPU texture, and then use that texture as a regular image that you can then scale in your game.

**Reply by grovesNL** [reply]
I'd probably render to a separate texture using a specific (smaller) font size as @BillyDM suggested. That will give you a lot of control over the scaling without having to change the internals of glyphon.

**Reply by BillyDM** [reply]
Another option could be to use a monospace bitmap font. Monospace fonts are easy enough to do your own layout (it's literally a grid). This is how retro video games worked.

**Reply by Patiga** [reply]
Thanks for the help with the thorough answers!

I suppose swash is smarter than I want/need and glyphon tightly integrates that.
I'll have to see if I can somehow manage to find a good solution with that.

Rendering indirectly via separate textures doesn't sound great performance-wise. The amount of texture allocations and render passes sounds like a bit too much, however over the last years I learned I don't have good intuition for performance in graphics programming. The game itself is very cheap to render, so I don't want to introduce new bottleneck with text rendering.

With the amount of strings I want to render eventually I'll kinda need to build a string atlas or some other data structure to keep strings in use.

I wonder at which point the sdf text rendering would be a viable option. I'd like to benchmark ofc, but these things tend to take a lot of time to implement, so I won't take the time upfront to implement two different options.
So far didn't find good resources yet which compare the two approaches, which I find curious.

**Reply by BillyDM** [reply]
Actually, rendering to a separate texture and then sampling from that texture in a separate render pass to transform it is quite performant (especially if you don't need to update the actual text very often). GPUs are designed for this kind of thing (it's how OS windowing compositors work).

**Reply by BillyDM** [reply]
More specifically, you only need to perform a render pass for text that has changed. Once the text has been rendered to a texture once, you can reuse that texture for future frames.

**Reply by BillyDM** [reply]
If you want to get fancy, you could even pack the rendered textures into a single texture to reduce texture switching overhead.

**Reply by BillyDM** [reply]
Oh actually, looking it up there is another solution, which is to tessellate the text into triangle meshes, and then just render the meshes on the GPU like normal. The epaint library from egui can do this: https://docs.rs/epaint/latest/epaint/tessellator/struct.Tessellator.html#method.tessellate_text

Though you might need to do MSAA to antialias the text.

**Reply by Patiga** [reply]
After reading up a bit on sdf font rendering it appears that their performance should actually be comparable.
Their drawback appears to be that it doesn't render the fonts perfectly due to rounded corners, which is a bit more complicated to fix.
Now I'm kinda motivated to write my own sdf wgpu text renderer. Like in kaku, single-sized text rendering can be provided as an alternative, as it is not much different.

`kaku` doesn't appear to use a texture atlas and there is some other stuff I want to do differently like allocations and dependencies, so I'll try to roll my own for a bit :)

**Reply by BillyDM** [reply]
Just use a rounded font :)
